homeassistant/components/airzone/binary_sensor.py

```diff
@@ -82,33 +82,50 @@ async def async_setup_entry(
     """Add Airzone binary sensors from a config_entry."""
     coordinator = entry.runtime_data
 
-    binary_sensors: list[AirzoneBinarySensor] = [
-        AirzoneSystemBinarySensor(
-            coordinator,
-            description,
-            entry,
-            system_id,
-            system_data,
-        )
-        for system_id, system_data in coordinator.data[AZD_SYSTEMS].items()
-        for description in SYSTEM_BINARY_SENSOR_TYPES
-        if description.key in system_data
-    ]
-
-    binary_sensors.extend(
-        AirzoneZoneBinarySensor(
-            coordinator,
-            description,
-            entry,
-            system_zone_id,
-            zone_data,
-        )
-        for system_zone_id, zone_data in coordinator.data[AZD_ZONES].items()
-        for description in ZONE_BINARY_SENSOR_TYPES
-        if description.key in zone_data
-    )
-
-    async_add_entities(binary_sensors)
+    added_systems: set[str] = set()
+    added_zones: set[str] = set()
+
+    def _async_entity_listener() -> None:
+        """Handle additions of binary sensors."""
+
+        systems_data = coordinator.data.get(AZD_SYSTEMS, {})
+        received_systems = set(systems_data)
+        new_systems = received_systems - added_systems
+        if new_systems:
+            async_add_entities(
+                AirzoneSystemBinarySensor(
+                    coordinator,
+                    description,
+                    entry,
+                    system_zone_id,
+                    systems_data.get(system_zone_id),
+                )
+                for system_zone_id in new_systems
+                for description in SYSTEM_BINARY_SENSOR_TYPES
+                if description.key in systems_data.get(system_zone_id)
+            )
+            added_systems.update(new_systems)
+
+        zones_data = coordinator.data.get(AZD_ZONES, {})
+        received_zones = set(zones_data)
+        new_zones = received_zones - added_zones
+        if new_zones:
+            async_add_entities(
```

Comment:
Can we combine the call? Async_add_entities is quite expensive iirc

Reply:
Done in f9c4dbb0f0b52442c7ef78ddb54e077c6cc62383